pkg/p2p/manager.go

```diff
@@ -0,0 +1,671 @@
+type PeerForEachFunc func(p *Peer) bool
+
+// ForEach calls the given PeerForEachFunc on each Peer.
+// Optionally only loops over the peers with the given filter relation.
+func (m *Manager) ForEach(f PeerForEachFunc, filter ...PeerRelation) {
+	back := make(chan struct{})
+	m.forEachChan <- &foreachmsg{f: f, back: back, filter: filter}
+	<-back
+}
+
+// ConnectedCount returns the count of connected peer.
+// Optionally only including peers with the given relation.
+func (m *Manager) ConnectedCount(relation ...PeerRelation) int {
+	var count int
+	m.ForEach(func(p *Peer) bool {
+		if m.host.Network().Connectedness(p.ID) == network.Connected {
+			count++
+		}
+		return true
+	}, relation...)
+	return count
+}
+
+// PeerSnapshots returns snapshots of information of peers known to the Manager.
+func (m *Manager) PeerSnapshots() []*PeerSnapshot {
+	infos := make([]*PeerSnapshot, 0)
+	m.ForEach(func(p *Peer) bool {
+		info := p.Info()
+		info.Connected = m.host.Network().Connectedness(p.ID) == network.Connected
+		infos = append(infos, info)
+		return true
+	})
+	return infos
+}
+
+// PeerFunc gets called with the given Peer.
+type PeerFunc func(p *Peer)
+
+// Call calls the given PeerFunc synchronized within the Manager's event loop, if the peer exists.
+// PeerFunc must not call any function on Manager.
+func (m *Manager) Call(id peer.ID, f PeerFunc) {
+	back := make(chan struct{})
+	m.callChan <- &callmsg{id: id, f: f, back: back}
+	<-back
+}
+
+type connectpeermsg struct {
+	addrInfo     *peer.AddrInfo
+	peerRelation PeerRelation
+	alias        string
+	back         chan error
+}
+
+type connectionmsg struct {
+	net  network.Network
+	conn network.Conn
+}
+
+type disconnectpeermsg struct {
+	id   peer.ID
+	back chan error
+}
+
+type isconnectedrequestmsg struct {
+	id   peer.ID
+	back chan bool
+}
+
+type reconnectmsg struct {
+	id peer.ID
+}
+
+type foreachmsg struct {
+	f      PeerForEachFunc
+	back   chan struct{}
+	filter []PeerRelation
+}
+
+type callmsg struct {
+	id   peer.ID
+	f    PeerFunc
+	back chan struct{}
+}
+
+// runs the Manager's event loop, we do operations on the Manager in this way,
+// because dealing with the natural concurrency of handling network connections
+// becomes very messy, especially since libp2p's notifiee system isn't clear on
+// what event is triggered when.
+func (m *Manager) eventLoop(shutdownSignal <-chan struct{}) {
+	for {
+		select {
+		case <-shutdownSignal:
+			return
+
+		case connectPeerMsg := <-m.connectPeerChan:
+			err := m.connectPeer(connectPeerMsg.addrInfo, connectPeerMsg.peerRelation, connectPeerMsg.alias)
+			if err != nil {
+				m.Events.Error.Trigger(fmt.Errorf("error connect to %s (%v): %w", connectPeerMsg.addrInfo.ID.ShortString(), connectPeerMsg.addrInfo.Addrs, err))
+			}
+			if errors.Is(err, ErrPeerInManagerAlready) {
+				m.updateRelation(connectPeerMsg.addrInfo.ID, connectPeerMsg.peerRelation)
+				m.updateAlias(connectPeerMsg.addrInfo.ID, connectPeerMsg.alias)
```

Comment:
Would that overwrite a known alias if a static peer with an alias gets connected via "autopeering" ?

Reply:
Yeah true it would. I'll add a check to not overwrite the alias in case it isn't empty.